File: logstash-core-event/lib/logstash/event.rb
# encoding: utf-8
require "time"
require "date"
require "cabin"
require "logstash/namespace"
require "logstash/util/accessors"
require "logstash/timestamp"
require "logstash/json"
require "logstash/string_interpolation"

# transcient pipeline events for normal in-flow signaling as opposed to
# flow altering exceptions. for now having base classes is adequate and
# in the future it might be necessary to refactor using like a BaseEvent
# class to have a common interface for all pileline events to support
# eventual queueing persistence for example, TBD.
class LogStash::ShutdownEvent; end
class LogStash::FlushEvent; end

module LogStash
  FLUSH = LogStash::FlushEvent.new

  # LogStash::SHUTDOWN is used by plugins
  SHUTDOWN = LogStash::ShutdownEvent.new
end

# the logstash event object.
#
# An event is simply a tuple of (timestamp, data).
# The 'timestamp' is an ISO8601 timestamp. Data is anything - any message,
# context, references, etc that are relevant to this event.
#
# Internally, this is represented as a hash with only two guaranteed fields.
#
# * "@timestamp" - an ISO8601 timestamp representing the time the event
#   occurred at.
# * "@version" - the version of the schema. Currently "1"
#
# They are prefixed with an "@" symbol to avoid clashing with your
# own custom fields.
#
# When serialized, this is represented in JSON. For example:
#
#     {
#       "@timestamp": "2013-02-09T20:39:26.234Z",
#       "@version": "1",
#       message: "hello world"
#     }
class LogStash::Event
  class DeprecatedMethod < StandardError; end

  CHAR_PLUS = "+"
  TIMESTAMP = "@timestamp"
  VERSION = "@version"
  VERSION_ONE = "1"
  TIMESTAMP_FAILURE_TAG = "_timestampparsefailure"
  TIMESTAMP_FAILURE_FIELD = "_@timestamp"

  METADATA = "@metadata".freeze
  METADATA_BRACKETS = "[#{METADATA}]".freeze

  # Floats outside of these upper and lower bounds are forcibly converted
  # to scientific notation by Float#to_s
  MIN_FLOAT_BEFORE_SCI_NOT = 0.0001
  MAX_FLOAT_BEFORE_SCI_NOT = 1000000000000000.0

  DEFAULT_LOGGER = Cabin::Channel.get(LogStash)
  @@logger = DEFAULT_LOGGER

  def initialize(data = {})
    @cancelled = false
    @data = data
    @accessors = LogStash::Util::Accessors.new(data)
    @data[VERSION] ||= VERSION_ONE
    ts = @data[TIMESTAMP]
    @data[TIMESTAMP] = ts ? init_timestamp(ts) : LogStash::Timestamp.now

    @metadata = @data.delete(METADATA) || {}
    @metadata_accessors = LogStash::Util::Accessors.new(@metadata)
  end

  def cancel
    @cancelled = true
  end

  def uncancel
    @cancelled = false
  end

  def cancelled?
    @cancelled
  end

  # Create a deep-ish copy of this event.
  def clone
    copy = {}
    @data.each do |k,v|
      # TODO(sissel): Recurse if this is a hash/array?
      copy[k] = begin v.clone rescue v end
    end

    self.class.new(copy)
  end

  def to_s
    "#{timestamp.to_iso8601} #{self.sprintf("%{host} %{message}")}"
  end

  def timestamp
    @data[TIMESTAMP]
  end

  def timestamp=(val)
    @data[TIMESTAMP] = val
  end

  def [](fieldref)
    if fieldref.start_with?(METADATA_BRACKETS)
      @metadata_accessors.get(fieldref[METADATA_BRACKETS.length .. -1])
    elsif fieldref == METADATA
      @metadata
    else
      @accessors.get(fieldref)
    end
  end

  def []=(fieldref, value)
    if fieldref == TIMESTAMP && !value.is_a?(LogStash::Timestamp)
      raise TypeError, "The field '@timestamp' must be a (LogStash::Timestamp, not a #{value.class} (#{value})"
    end
    if fieldref.start_with?(METADATA_BRACKETS)
      @metadata_accessors.set(fieldref[METADATA_BRACKETS.length .. -1], value)
    elsif fieldref == METADATA
      @metadata = value
      @metadata_accessors = LogStash::Util::Accessors.new(@metadata)
    else
      @accessors.set(fieldref, value)
    end
  end

  def to_json(*args)
    # ignore arguments to respect accepted to_json method signature
    LogStash::Json.dump(@data)
  end

  def to_hash
    @data
  end

  def overwrite(event)
    # pickup new event @data and also pickup @accessors
    # otherwise it will be pointing on previous data
    @data = event.instance_variable_get(:@data)
    @accessors = event.instance_variable_get(:@accessors)

    #convert timestamp if it is a String
    if @data[TIMESTAMP].is_a?(String)
      @data[TIMESTAMP] = LogStash::Timestamp.parse_iso8601(@data[TIMESTAMP])
    end
  end

  def include?(fieldref)
    if fieldref.start_with?(METADATA_BRACKETS)
      @metadata_accessors.include?(fieldref[METADATA_BRACKETS.length .. -1])
    elsif fieldref == METADATA
      true
    else
      @accessors.include?(fieldref)
    end
  end

  # Append an event to this one.
  def append(event)
    # non-destructively merge that event with ourselves.

    # no need to reset @accessors here because merging will not disrupt any existing field paths
    # and if new ones are created they will be picked up.
    LogStash::Util.hash_merge(@data, event.to_hash)
  end

  # Remove a field or field reference. Returns the value of that field when deleted
  def remove(fieldref)
    @accessors.del(fieldref)
  end

  # sprintf. This could use a better method name.
  # The idea is to take an event and convert it to a string based on
  # any format values, delimited by %{foo} where 'foo' is a field or
  # metadata member.
  #
  # For example, if the event has type == "foo" and host == "bar"
  # then this string:
  #   "type is %{type} and source is %{host}"
  # will return
  #   "type is foo and source is bar"
  #
  # If a %{name} value is an array, then we will join by ','
  # If a %{name} value does not exist, then no substitution occurs.
  def sprintf(format)
    LogStash::StringInterpolation.evaluate(self, format)
  end

  def tag(value)
    # Generalize this method for more usability
    self["tags"] ||= []
    self["tags"] << value unless self["tags"].include?(value)
  end

  def to_hash_with_metadata
    @metadata.empty? ? to_hash : to_hash.merge(METADATA => @metadata)
  end

  def to_json_with_metadata(*args)
    # ignore arguments to respect accepted to_json method signature
    LogStash::Json.dump(to_hash_with_metadata)
  end

  # this is used by logstash-devutils spec_helper.rb to monkey patch the Event field setter []=
  # and add systematic encoding validation on every field set in specs.
  # TODO: (colin) this should be moved, probably in logstash-devutils ?
  def self.validate_value(value)
    case value
    when String
      raise("expected UTF-8 encoding for value=#{value}, encoding=#{value.encoding.inspect}") unless value.encoding == Encoding::UTF_8
      raise("invalid UTF-8 encoding for value=#{value}, encoding=#{value.encoding.inspect}") unless value.valid_encoding?
      value
    when Array
      value.each{|v| validate_value(v)} # don't map, return original object
      value
    else
      value
    end
  end

  # depracated public methods
  # TODO: (colin) since these depracated mothods are still exposed in 2.x we should remove them in 3.0

  def unix_timestamp
    raise DeprecatedMethod
  end

  def ruby_timestamp
    raise DeprecatedMethod
  end

  def fields
    raise DeprecatedMethod
  end

  # set a new logger for all Event instances
  # there is no point in changing it at runtime for other reasons than in tests/specs.
  # @param logger [Cabin::Channel] logger instance that will be used by all Event instances
  def self.logger=(logger)
    @@logger = logger
  end

  private

  def logger
    @@logger
  end

  def init_timestamp(o)
    begin
      timestamp = LogStash::Timestamp.coerce(o)
      return timestamp if timestamp

      logger.warn("Unrecognized #{TIMESTAMP} value, setting current time to #{TIMESTAMP}, original in #{TIMESTAMP_FAILURE_FIELD}field", :value => o.inspect)
    rescue LogStash::TimestampParserError => e
      logger.warn("Error parsing #{TIMESTAMP} string, setting current time to #{TIMESTAMP}, original in #{TIMESTAMP_FAILURE_FIELD} field", :value => o.inspect, :exception => e.message)
    end

    @data["tags"] ||= []
    @data["tags"] << TIMESTAMP_FAILURE_TAG unless @data["tags"].include?(TIMESTAMP_FAILURE_TAG)
    @data[TIMESTAMP_FAILURE_FIELD] = o

    LogStash::Timestamp.now
  end
end

File: logstash-core-event-java/src/main/java/com/logstash/Event.java
package com.logstash;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.logstash.bivalues.NullBiValue;
import com.logstash.bivalues.StringBiValue;
import com.logstash.bivalues.TimeBiValue;
import com.logstash.bivalues.TimestampBiValue;
import com.logstash.ext.JrubyTimestampExtLibrary;
import org.joda.time.DateTime;
import org.jruby.RubySymbol;

import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class Event implements Cloneable, Serializable {

    private boolean cancelled;
    private Map<String, Object> data;
    private Map<String, Object> metadata;
    private Timestamp timestamp;
    private Accessors accessors;
    private Accessors metadata_accessors;

    public static final String METADATA = "@metadata";
    public static final String METADATA_BRACKETS = "[" + METADATA + "]";
    public static final String TIMESTAMP = "@timestamp";
    public static final String TIMESTAMP_FAILURE_TAG = "_timestampparsefailure";
    public static final String TIMESTAMP_FAILURE_FIELD = "_@timestamp";
    public static final String VERSION = "@version";
    public static final String VERSION_ONE = "1";

    private static final Logger DEFAULT_LOGGER = new StdioLogger();
    private static final ObjectMapper mapper = new ObjectMapper();

    // logger is static since once set there is no point in changing it at runtime
    // for other reasons than in tests/specs.
    private transient static Logger logger = DEFAULT_LOGGER;

    public Event()
    {
        this.metadata = new HashMap<String, Object>();
        this.data = new HashMap<String, Object>();
        this.data.put(VERSION, VERSION_ONE);
        this.cancelled = false;
        this.timestamp = new Timestamp();
        this.data.put(TIMESTAMP, this.timestamp);
        this.accessors = new Accessors(this.data);
        this.metadata_accessors = new Accessors(this.metadata);
    }

    public Event(Map data)
    {
        this.data = (Map<String, Object>)Valuefier.convert(data);

        if (!this.data.containsKey(VERSION)) {
            this.data.put(VERSION, VERSION_ONE);
        }

        if (this.data.containsKey(METADATA)) {
            this.metadata = (Map<String, Object>) this.data.remove(METADATA);
        } else {
            this.metadata = new HashMap<String, Object>();
        }
        this.metadata_accessors = new Accessors(this.metadata);

        this.cancelled = false;
        this.timestamp = initTimestamp(data.get(TIMESTAMP));
        this.data.put(TIMESTAMP, this.timestamp);
        this.accessors = new Accessors(this.data);
    }

    public Map<String, Object> getData() {
        return this.data;
    }

    public Map<String, Object> getMetadata() {
        return this.metadata;
    }

    public void setData(Map<String, Object> data) {
        this.data = ConvertedMap.newFromMap(data);
    }

    public Accessors getAccessors() {
        return this.accessors;
    }

    public Accessors getMetadataAccessors() {
        return this.metadata_accessors;
    }

    public void setAccessors(Accessors accessors) {
        this.accessors = accessors;
    }

    public void setMetadataAccessors(Accessors accessors) {
        this.metadata_accessors = accessors;
    }

    public void cancel() {
        this.cancelled = true;
    }

    public void uncancel() {
        this.cancelled = false;
    }

    public boolean isCancelled() {
        return this.cancelled;
    }

    public Timestamp getTimestamp() throws IOException {
        if (this.data.containsKey(TIMESTAMP)) {
            return this.timestamp;
        } else {
            throw new IOException("fails");
        }
    }

    public void setTimestamp(Timestamp t) {
        this.timestamp = t;
        this.data.put(TIMESTAMP, this.timestamp);
    }

    public Object getField(String reference) {
        Object val = getUnconvertedField(reference);
        return Javafier.deep(val);
    }

    public Object getUnconvertedField(String reference) {
        if (reference.equals(METADATA)) {
            return this.metadata;
        } else if (reference.startsWith(METADATA_BRACKETS)) {
            return this.metadata_accessors.get(reference.substring(METADATA_BRACKETS.length()));
        } else {
            return this.accessors.get(reference);
        }
    }

    public void setField(String reference, Object value) {
        if (reference.equals(TIMESTAMP)) {
            // TODO(talevy): check type of timestamp
            this.accessors.set(reference, value);
        } else if (reference.equals(METADATA_BRACKETS) || reference.equals(METADATA)) {
            this.metadata = (Map<String, Object>) value;
            this.metadata_accessors = new Accessors(this.metadata);
        } else if (reference.startsWith(METADATA_BRACKETS)) {
            this.metadata_accessors.set(reference.substring(METADATA_BRACKETS.length()), value);
        } else {
            this.accessors.set(reference, Valuefier.convert(value));
        }
    }

    public boolean includes(String reference) {
        if (reference.equals(METADATA_BRACKETS) || reference.equals(METADATA)) {
            return true;
        } else if (reference.startsWith(METADATA_BRACKETS)) {
            return this.metadata_accessors.includes(reference.substring(METADATA_BRACKETS.length()));
        } else {
            return this.accessors.includes(reference);
        }
    }

    public String toJson()
            throws IOException
    {
        return mapper.writeValueAsString(this.data);
    }

    public static Event[] fromJson(String json)
            throws IOException
    {
        // empty/blank json string does not generate an event
        if (json == null || json.trim().isEmpty()) {
            return new Event[]{ };
        }

        Event[] result;
        Object o = mapper.readValue(json, Object.class);
        // we currently only support Map or Array json objects
        if (o instanceof Map) {
            result = new Event[]{ new Event((Map)o) };
        } else if (o instanceof List) {
            result = new Event[((List) o).size()];
            int i = 0;
            for (Object e : (List)o) {
                if (!(e instanceof Map)) {
                    throw new IOException("incompatible inner json array object type=" + e.getClass().getName() + " , only hash map is suppoted");
                }
                result[i++] = new Event((Map)e);
            }
        } else {
            throw new IOException("incompatible json object type=" + o.getClass().getName() + " , only hash map or arrays are suppoted");
        }

        return result;
    }

    public Map toMap() {
        return Cloner.deep(this.data);
    }

    public Event overwrite(Event e) {
        this.data = e.getData();
        this.accessors = e.getAccessors();
        this.cancelled = e.isCancelled();
        try {
            this.timestamp = e.getTimestamp();
        } catch (IOException exception) {
            this.timestamp = new Timestamp();
        }

        return this;
    }

    public Event append(Event e) {
        Util.mapMerge(this.data, e.data);

        return this;
    }

    public Object remove(String path) {
        return this.accessors.del(path);
    }

    public String sprintf(String s) throws IOException {
        return StringInterpolation.getInstance().evaluate(this, s);
    }

    public Event clone()
            throws CloneNotSupportedException
    {
//        Event clone = (Event)super.clone();
//        clone.setAccessors(new Accessors(clone.getData()));

        Event clone = new Event(Cloner.deep(getData()));
        return clone;
    }

    public String toString() {
        // TODO: (colin) clean this IOException handling, not sure why we bubble IOException here
        try {
            return (getTimestamp().toIso8601() + " " + this.sprintf("%{host} %{message}"));
        } catch (IOException e) {
            String host = (String)this.data.get("host");
            host = (host != null ? host : "%{host}");

            String message = (String)this.data.get("message");
            message = (message != null ? message : "%{message}");

            return (host + " " + message);
        }
    }

    private Timestamp initTimestamp(Object o) {
        try {
            if (o == null || o instanceof NullBiValue) {
                // most frequent
                return new Timestamp();
            } else if (o instanceof String) {
                // second most frequent
                return new Timestamp((String) o);
            } else if (o instanceof StringBiValue) {
                return new Timestamp(((StringBiValue) o).javaValue());
            } else if (o instanceof TimeBiValue) {
                return new Timestamp(((TimeBiValue) o).javaValue());
            } else if (o instanceof JrubyTimestampExtLibrary.RubyTimestamp) {
                return new Timestamp(((JrubyTimestampExtLibrary.RubyTimestamp) o).getTimestamp());
            } else if (o instanceof Timestamp) {
                return new Timestamp((Timestamp) o);
            } else if (o instanceof TimestampBiValue) {
                return new Timestamp(((TimestampBiValue) o).javaValue());
            } else if (o instanceof DateTime) {
                return new Timestamp((DateTime) o);
            } else if (o instanceof Date) {
                return new Timestamp((Date) o);
            } else if (o instanceof RubySymbol) {
                return new Timestamp(((RubySymbol) o).asJavaString());
            } else {
                Event.logger.warn("Unrecognized " + TIMESTAMP + " value type=" + o.getClass().toString());
            }
        } catch (IllegalArgumentException e) {
            Event.logger.warn("Error parsing " + TIMESTAMP + " string value=" + o.toString());
        }

        tag(TIMESTAMP_FAILURE_TAG);
        this.data.put(TIMESTAMP_FAILURE_FIELD, o);

        return Timestamp.now();
    }

    public void tag(String tag) {
        List<Object> tags = (List<Object>) this.data.get("tags");
        if (tags == null) {
            tags = new ArrayList<>();
            this.data.put("tags", tags);
        }

        if (!tags.contains(tag)) {
            tags.add(tag);
        }
    }

    // Event.logger is static since once set there is no point in changing it at runtime
    // for other reasons than in tests/specs.
    public static void setLogger(Logger logger) {
        Event.logger = logger;
    }
}
